### PR TITLE
Add static analysis for unsafe uint casting

### PR DESCRIPTION
**Pull request by prestonvanloon**
**What type of PR is this?**

Other

**What does this PR do? Why is it needed?**

Casting uint types to int types can lead to unexpected results, i.e. overflows.

This PR adds a static analysis check for these scenarios and resolves all found issues.

**Which issues(s) does this PR fix?**

**Other notes for review**

One thing I thought about is that it's technically not a big deal to cast a smaller uint to a larger int (uint32 -> int64), but we typically only deal with uint64 so I decided to warn on any conversion from uint to int.

### CODE REVIEW

**Comment by rauljordan** [comment]
Note this closes https://github.com/prysmaticlabs/prysm/issues/9958 if all exceptions are taken care of

File: beacon-chain/p2p/encoder/ssz.go
Comment (prestonvanloon) [inline beacon-chain/p2p/encoder/ssz.go:148]: This validation has moved to math.Int

File: proto/prysm/v1alpha1/attestation/aggregation/testing/BUILD.bazel
Comment (prestonvanloon) [inline proto/prysm/v1alpha1/attestation/aggregation/testing/BUILD.bazel:5]: Found this in my journey. More test code not marked as testonly :(

File: WORKSPACE
Comment (rauljordan) [inline WORKSPACE:399]: Should be in deps.bzl right?

File: beacon-chain/forkchoice/protoarray/helpers.go
Comment (rauljordan) [inline beacon-chain/forkchoice/protoarray/helpers.go:66]: Potuz and I had a comment here about casting the balances to ints when tehy should probably have remained as uint64

File: tools/analyzers/uintcast/analyzer_test.go
Comment (rauljordan) [inline tools/analyzers/uintcast/analyzer_test.go:10]: Why is this test excluded from Bazel?

**Review by rauljordan — COMMENTED** [review]
Found a test case the analyzer does not catch: when doing the cast during struct initialization:

```go
func Uint64CastInStruct() {
	type S struct {
		a int
	}
	s := S{
		a: int(uint64(5)), // want "Unsafe cast from uint64 to int."
	}
	_ = s
}
```
this fails when running the Analyzer test

File: WORKSPACE
Comment (prestonvanloon) [reply WORKSPACE:399]: Yes, oops!

File: tools/analyzers/uintcast/analyzer_test.go
Comment (prestonvanloon) [reply tools/analyzers/uintcast/analyzer_test.go:10]: It doesn't work. It depends on $GOROOT. I left a TODO, ill try to fix it

File: beacon-chain/cache/active_balance.go
Comment (nisdas) [reply beacon-chain/cache/active_balance.go:23]: Ah k, I see it is due to the `lruwrpr`, maybe we can change the method signature to take a `uint64`. And it does the safe casting internally in the lru constructor.

Comment (prestonvanloon) [reply beacon-chain/cache/active_balance.go:23]: Yeah exactly, I moved these to `const` so that we are assured it can never change. I thought this was the easier route vs changing 3rd party API

Comment (prestonvanloon) [reply beacon-chain/cache/active_balance.go:23]: Oh! OK, I'll take a look!

Comment (prestonvanloon) [reply beacon-chain/cache/active_balance.go:23]: Why do we even have `lruwrpr`? It's a flow through to ` github.com/hashicorp/golang-lru` with panics. I'm going to leave it out of scope for this PR. It seems strange to have this package when we could use lru directly.

File: beacon-chain/forkchoice/protoarray/helpers.go
Comment (prestonvanloon) [reply beacon-chain/forkchoice/protoarray/helpers.go:66]: Discussed this offline. The balances must be cast to int as the delta could be negative.

File: tools/analyzers/uintcast/analyzer_test.go
Comment (prestonvanloon) [reply tools/analyzers/uintcast/analyzer_test.go:17]: OK. I fixed this and then fixed all of the analyzer tests.

**Comment by prestonvanloon** [comment]
> > Found a test case the analyzer does not catch: when doing the cast during struct initialization:
> > ```go
> > func Uint64CastInStruct() {
> > 	type S struct {
> > 		a int
> > 	}
> > 	s := S{
> > 		a: int(uint64(5)), // want "Unsafe cast from uint64 to int."
> > 	}
> > 	_ = s
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > this fails when running the Analyzer test
> 
> This is a pretty contrived example. I will see if I can reasonably add to the SA to solve this, but I might not be able to do it in this PR

Thinking about this example more. It is not as contrived as i thought. For example, I think a function call with a single return type of `uint` would also fail to be captured in the static analysis. I'll revisit this idea today.

**Comment by prestonvanloon** [comment]
> ```go
> func Uint64CastInStruct() {
> 	type S struct {
> 		a int
> 	}
> 	s := S{
> 		a: int(uint64(5)), // want "Unsafe cast from uint64 to int."
> 	}
> 	_ = s
> }
> ```

```go
func Uint64CastInStruct() {
	type S struct {
		a int
	}
	s := S{
		a: int(uint64(5)), // want "Unsafe cast from uint64 to int."
	}
	_ = s
}

func Uint64CastFunctionReturn() {
	fn := func() uint64 {
		return 5
	}
	a := int(fn()) // want "Unsafe cast from uint64 to int."
	_ = a
}
```

Both of these fail. The former test is pretty important actually...

**Review by uncdr — COMMENTED** [review]
i'm still not ok where int, uint, uint64 should be used

File: beacon-chain/cache/committee.go
Comment (prestonvanloon) [reply beacon-chain/cache/committee.go:26]: That's the 3rd party API. I'm not sure. This is a constant now so it can't overflow/underflow. I think we can address cache API in another PR or issue.

File: beacon-chain/p2p/message_id.go
Comment (prestonvanloon) [reply beacon-chain/p2p/message_id.go:95]: I'll add a SafeAddInt64 method for this

File: beacon-chain/cache/active_balance.go
Comment (nisdas) [reply beacon-chain/cache/active_balance.go:23]: I think we wanted the lru package to not return an error mostly. The package returns an error if the cache size is negative.

**Comment by prestonvanloon** [comment]
Removing OK to merge until after someone approves this. Since this PR adds new static analysis, there is no cache hits between develop based branches and this one. So this PR is contributing to high build agent disk usage until it can be merged.

**Review by nisdas — APPROVED** [review]

**Pull request merged**
